### DISCUSSION TITLE
Listening to a text controller inside Riverpod

### DISCUSSION THREAD

**Discussion by 2kjm** [rrousselGit/riverpod#1440] [Q&A] [answered]
I want to listen to a textcontroller and dispatch an action if it's empty. While I try to implement inside initstate i get a Badstate error. What is the right way to do this?

**Comment by msarkrish** [comment]
@2kjm  Can you post the error you have received?

**Reply by 2kjm** [reply]
`Another exception was thrown: Bad state: Tried to read a provider from a ProviderContainer that was already disposed`

**Reply by msarkrish** [reply]
Can you post your sample code?

**Comment by mark8044** [comment, marked as answer]
Without some code its hard to say what the exact issue is, but my goto general solution for state situations is to wrap a ref.read inside a `postFrameCallBack`, for better or worse

```
                WidgetsBinding.instance.addPostFrameCallback((timeStamp) {
                              ref.read(provider)....
                    });
```

**Reply by 2kjm** [reply]
I want to dispatch like this 


`  @override
  void initState() {
    widget.textController.addListener(() {
      if (widget.textController.text.isEmpty) {
        ref.watch(messageComposerProvider.notifier).setIsComposing(false);
      }
    });
    super.initState();
  }`

**Reply by tbm98** [reply]
use ref.read instead of watch

**Reply by 2kjm** [reply]
Tried it with ref.read but still getting error

my code 
`  @override
  void initState() {
    widget.textController.addListener(() {
      if (widget.textController.text.isEmpty) {
        ref.read(messageComposerProvider.notifier).setIsComposing(false);
      } else {
        ref
            .read(messageComposerProvider.notifier)
            .setNoOfLinesInTextField(widget.textController.text);
        ref.read(messageComposerProvider.notifier).setIsComposing(true);
      }    
    });
    super.initState();
  }`

my error:


`══╡ EXCEPTION CAUGHT BY FOUNDATION LIBRARY ╞════════════════════════════════════════════════════════
The following assertion was thrown while dispatching notifications for TextEditingController:
Looking up a deactivated widget's ancestor is unsafe.
At this point the state of the widget's element tree is no longer stable.
To safely refer to a widget's ancestor in its dispose() method, save a reference to the ancestor by
calling dependOnInheritedWidgetOfExactType() in the widget's didChangeDependencies() method.

When the exception was thrown, this was the stack:
#0      Element._debugCheckStateIsActiveForAncestorLookup.<anonymous closure> (package:flutter/src/widgets/framework.dart:4186:9)
#1      Element._debugCheckStateIsActiveForAncestorLookup (package:flutter/src/widgets/framework.dart:4200:6)
#2      Element.getElementForInheritedWidgetOfExactType (package:flutter/src/widgets/framework.dart:4226:12)
#3      ProviderScope.containerOf (package:flutter_riverpod/src/framework.dart:97:12)
#4      ConsumerStatefulElement.read (package:flutter_riverpod/src/consumer.dart:487:26)
#5      _TextMessageComposerState.initState.<anonymous closure> (package:message_composer/text_message_composer.dart:22:13)
#6      ChangeNotifier.notifyListeners (package:flutter/src/foundation/change_notifier.dart:324:24)
#7      ValueNotifier.value= (package:flutter/src/foundation/change_notifier.dart:428:5)
#8      TextEditingController.value= (package:flutter/src/widgets/editable_text.dart:159:11)
#9      EditableTextState._value= (package:flutter/src/widgets/editable_text.dart:2211:23)
#10     EditableTextState._formatAndSetValue (package:flutter/src/widgets/editable_text.dart:2617:5)
#11     EditableTextState.userUpdateTextEditingValue (package:flutter/src/widgets/editable_text.dart:2919:5)
#12     RenderEditable._setTextEditingValue (package:flutter/src/rendering/editable.dart:710:27)
#13     RenderEditable._setSelection (package:flutter/src/rendering/editable.dart:729:5)
#14     RenderEditable.selectWordEdge (package:flutter/src/rendering/editable.dart:2011:5)
#15     TextSelectionGestureDetectorBuilder.onSingleTapUp (package:flutter/src/widgets/text_selection.dart:1539:30)
#16     _TextFieldSelectionGestureDetectorBuilder.onSingleTapUp (package:flutter/src/material/text_field.dart:89:11)
#17     _TextSelectionGestureDetectorState._handleTapUp (package:flutter/src/widgets/text_selection.dart:1966:29)
#18     TapGestureRecognizer.handleTapUp.<anonymous closure> (package:flutter/src/gestures/tap.dart:611:57)
#19     GestureRecognizer.invokeCallback (package:flutter/src/gestures/recognizer.dart:198:24)
#20     TapGestureRecognizer.handleTapUp (package:flutter/src/gestures/tap.dart:611:11)
#21     BaseTapGestureRecognizer._checkUp (package:flutter/src/gestures/tap.dart:298:5)
#22     BaseTapGestureRecognizer.acceptGesture (package:flutter/src/gestures/tap.dart:269:7)
#23     GestureArenaManager.sweep (package:flutter/src/gestures/arena.dart:157:27)
#24     GestureBinding.handleEvent (package:flutter/src/gestures/binding.dart:449:20)
#25     GestureBinding.dispatchEvent (package:flutter/src/gestures/binding.dart:425:22)
#26     RendererBinding.dispatchEvent (package:flutter/src/rendering/binding.dart:329:11)
#27     GestureBinding._handlePointerEventImmediately (package:flutter/src/gestures/binding.dart:380:7)
#28     GestureBinding.handlePointerEvent (package:flutter/src/gestures/binding.dart:344:5)
#29     GestureBinding._flushPointerEventQueue (package:flutter/src/gestures/binding.dart:302:7)
#30     GestureBinding._handlePointerDataPacket (package:flutter/src/gestures/binding.dart:285:7)
#34     _invoke1 (dart:ui/hooks.dart:170:10)
#35     PlatformDispatcher._dispatchPointerDataPacket (dart:ui/platform_dispatcher.dart:331:7)
#36     _dispatchPointerDataPacket (dart:ui/hooks.dart:94:31)
(elided 3 frames from dart:async)

The TextEditingController sending notification was:
  TextEditingController#d2b7c(TextEditingValue(text: ┤├, selection: TextSelection.collapsed(offset:
  0, affinity: TextAffinity.downstream, isDirectional: false), composing: TextRange(start: -1, end:
  -1)))
════════════════════════════════════════════════════════════════════════════════════════════════════`

**Reply by mark8044** [reply]
@2kjm Why not try what I mentioned and put it inside a `postFrameCallback`?

```
@override void initState() { 
widget.textController.addListener(() { 
if (widget.textController.text.isEmpty) 
{ 
WidgetsBinding.instance.addPostFrameCallback((timeStamp) {
                              ref.read(messageComposerProvider.notifier).setIsComposing(false); 
                    });
} 
else 
{ 
WidgetsBinding.instance.addPostFrameCallback((timeStamp) {
ref .read(messageComposerProvider.notifier) .setNoOfLinesInTextField(widget.textController.text); ref.read(messageComposerProvider.notifier).setIsComposing(true); 
                    });
} 
}); 
super.initState(); 
}
```

**Reply by 2kjm** [reply]
Thanks man. It worked

**Reply by XuanTung95** [reply]
@2kjm 

```
The following assertion was thrown while dispatching notifications for TextEditingController:
Looking up a deactivated widget's ancestor is unsafe.
At this point the state of the widget's element tree is no longer stable.
To safely refer to a widget's ancestor in its dispose() method, save a reference to the ancestor by
```

I think this exception occurs because you use textController.addListener but do not remove it, and textController remain after the ConsumerState is disposed, hence the old listener causes the exception. I think creating TextEditingController outside of State and passing it around can cause hidden bugs, should avoid.
